src/cargo/util/toml/mod.rs

```diff
@@ -104,13 +104,41 @@ fn read_manifest_from_str(
     let package_root = manifest_file.parent().unwrap();
 
     let mut unused = BTreeSet::new();
+    let mut public_deps_without_z = BTreeSet::new();
+
     let deserializer = toml::de::Deserializer::new(contents);
     let manifest: manifest::TomlManifest = match serde_ignored::deserialize(deserializer, |path| {
         let mut key = String::new();
         stringify(&mut key, &path);
         unused.insert(key);
     }) {
-        Ok(manifest) => manifest,
+        Ok(manifest) => {
+            let mut manifest: manifest::TomlManifest = manifest;
+            let is_public_dependency_enabled = {
```

Comment:
If no `public-dependency` and `-Zpublic-dependency`,  set public to None.
Then will warn people and encourage people to `-Zpublic-dependency`.

This behavior had test case to verify.